File: source/_posts/高级实现/一致性哈希算法实现/ConsistHashing.php
<?php
class ConsistHashingHelper {
    //服务器信息
    private $SERVERS = array("node1", "node2", "node3","node4", "node5", "node6");
    //一致性哈希环
    private $hashRing = array();
    //虚拟节点数目
    const VIRTUAL_NODE_COUNT = 32;
    //hash环的最大位置2^32
    const MAX_HASH_LOCATION = 1<<32;
    public  $temp = '';
    public function __construct()
    {
        $this->initHashRingWithVirtualNode();
    }
    //初始化哈希环，没有虚拟节点
    public function initHashRingWithoutVirtualNode()
    {
        foreach($this->SERVERS as $server) {
            $this->hashRing[self::hash($server)] = $server;
        }
    }
    //初始化哈希环，没有虚拟节点
    public function initHashRingWithVirtualNode()
    {
        foreach($this->SERVERS as $server) {
            for ($i=0; $i<self::VIRTUAL_NODE_COUNT; $i++) {
                $virtualServer = $server ."#virtual" .$i;
                $this->temp .= ','.$virtualServer;
                $this->hashRing[self::hash($virtualServer)] = $virtualServer;
            }
        }
    }
    //计算hash值,常用crc32算法
    public static function hash($key) {
        $crc32Int = crc32($key);
        $crc32Int = abs($crc32Int);
        return $crc32Int % self::MAX_HASH_LOCATION;
    }
    //根据虚节点获取对应物理节点
    public function getRealNode($virtualNode) {
        if (strpos($virtualNode, '#') === false) {
            return $virtualNode;
        }
        return explode('#', $virtualNode)[0];
    }
    //为key分配server(真实物理节点)
    public  function allocateServer($key) {
        $hashValue = self::hash($key);
        $tailHashRing = $this->getNextAll($hashValue);
        var_dump(count($tailHashRing));
        if (empty($tailHashRing)) {
            reset($this->hashRing);
            $virtualNode = current($this->hashRing);
        } else {
            $virtualNode = current($tailHashRing);
        }
        return $this->getRealNode($virtualNode);
    }
    //获取对应的虚节点
    public  function  getNextAll($findKey){
        $flag = false; // 标识是否找到了键相等的元素即环上的节点
        $tailHashRing = array();//找到的节点及其后面的所有节点
        foreach ($this->hashRing as $key => $val) {
            if ($findKey == $key) {
                $flag = true;
                $tailHashRing[$key] = $val;
            } else {
                if ($flag == true) {
                    $tailHashRing[$key] = $val;
                }
            }
        }
        return $tailHashRing;
    }
    public function getHashRing() {
        return $this->hashRing;
    }
}
$consistHashingHelper = new ConsistHashingHelper();
/*foreach (array_keys($consistHashingHelper->getHashRing()) as $key) {
    echo $key.PHP_EOL;
}*/
$node1Count = 0;
$node2Count = 0;
$node3Count = 0;
$node4Count = 0;
$node5Count = 0;
$node6Count = 0;
$str = $consistHashingHelper->temp;
var_dump($str);
$data = explode(',', $str);
$data = array_filter($data, function ($val) {
    if (empty($val)){
        return false;
    } else {
        return true;
    }
});
$data = array_values($data);
for ($i=0; $i<count($data);$i++) {
    $key = $data[$i];
    $server = $consistHashingHelper->allocateServer($key);
    if (strpos($server, 'node1') !== false ) {
        $node1Count++;
    } else if (strpos($server, 'node2') !== false ) {
        $node2Count++;
    } else if (strpos($server, 'node3') !== false ) {
        $node3Count++;
    } else if (strpos($server, 'node4') !== false ) {
        $node4Count++;
    }else if (strpos($server, 'node5') !== false ) {
        $node5Count++;
    }else if (strpos($server, 'node6') !== false ) {
        $node6Count++;
    }else {
        echo '没有找到匹配项'.PHP_EOL;
    }
}
echo "node1Count=" . $node1Count.PHP_EOL;
echo "node2Count=" . $node2Count.PHP_EOL;
echo "node3Count=" . $node3Count.PHP_EOL;
echo "node4Count=" . $node4Count.PHP_EOL;
echo "node5Count=" . $node5Count.PHP_EOL;
echo "node6Count=" . $node6Count.PHP_EOL;

File: source/_posts/高级实现/一致性哈希算法实现/ConsistHashing.java
package Id;
import java.util.Iterator;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.UUID;
import java.util.zip.CRC32;
public class ConsistHashing {
	/**
     * 服务器信息
     */
    private static final String[] SERVERS = new String[]{"node1", "node2", "node3"};

    /**
     * 一致性哈希环
     */
    private static SortedMap<Long, String> hashRing = new TreeMap<>();

    /**
     * 虚拟节点数目
     */
    private static final int VIRTUAL_NODE_COUNT = 32;

    /**
     * hash环的最大位置2^32
     */
    private static final long MAX_HASH_LOCATION = 1L << 32;

    static {
        //initHashRingWithoutVirtualNode();
        initHashRingWithVirtualNode();
    }

    /**
     * 初始化哈希环，没有虚拟节点
     */
    public static void initHashRingWithoutVirtualNode() {
        for (String server : SERVERS) {
            hashRing.put(hash(server), server);
        }
    }

    /**
     * 初始化哈希环，带虚拟节点
     */
    public static void initHashRingWithVirtualNode() {
        for (String server : SERVERS) {
            for (int i = 0; i < VIRTUAL_NODE_COUNT; i++) {
                String virtualServer = server + "#virtual" + i;
                hashRing.put(hash(virtualServer), virtualServer);
            }
        }
    }

    /**
     * 为key分配server
     *
     * @param key key
     * @return 分配的server
     */
    public static String allocateServer(String key) {
        long hashValue = hash(key);
        /*
        * TreeMap保证key的有序排列，可以当作是一个首尾相连的环,
        * 所以为新的key分配server地址时，按照相同的hash算法得到hashValue，
        * 在TreeMap的环状结构中，查找第一个比该hashValue大的key，其对应的server即为分配的server
        * */
        SortedMap<Long, String> tailHashRing = hashRing.tailMap(hashValue);
        String virtualNode;
        if (tailHashRing.isEmpty()) {
            virtualNode = hashRing.get(hashRing.firstKey());
        } else {
            virtualNode = hashRing.get(tailHashRing.firstKey());
        }
        return getRealNode(virtualNode);
    }


    private static long hash(String key) {
        CRC32 crc32 = new CRC32();
        crc32.update(key.getBytes());
        System.out.println(crc32.getValue());
        return crc32.getValue() % MAX_HASH_LOCATION;
    }

    private static String getRealNode(String virtualNode) {
        if (!virtualNode.contains("#")) {
            return virtualNode;
        }
        return virtualNode.split("#")[0];
    }

	public static void main(String[] args) {
		Iterator<Long> iterator = hashRing.keySet().iterator();
        while (iterator.hasNext()) {
            Long key = iterator.next();
            System.out.println(key + "," + hashRing.get(key));
        }
        int node1Count = 0;
        int node2Count = 0;
        int node3Count = 0;

        for (int i = 0; i < 1000; i++) {
            String key = UUID.randomUUID().toString();
            System.out.println(key);
            String server = allocateServer(key);
            if (server.equals("node1")) node1Count++;
            if (server.equals("node2")) node2Count++;
            if (server.equals("node3")) node3Count++;
        }
        System.out.println("node1Count=" + node1Count);
        System.out.println("node2Count=" + node2Count);
        System.out.println("node3Count=" + node3Count);

	}

}
